[Media][MSE] Check that the media element can still be used after a decoding error

eocanha opened the pull request:
#### 02759542825a229e6c4284abf3cdbcab25f78bac
<pre>
[Media][MSE] Check that the media element can still be used after a decoding error
<a href="https://bugs.webkit.org/show_bug.cgi?id=242384">https://bugs.webkit.org/show_bug.cgi?id=242384</a>

Reviewed by Xabier Rodriguez-Calvar.

The media source instance should be detached before destroying MSE private player. Otherwise, it can create a
crash under some circumstances that can only properly be reproduced in downstream ports with specific devices.

See: <a href="https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/890">https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/890</a>

A test to check that the player doesn&apos;t crash when an MSE playback decoding error happens has been added.
This test uses a corrupt AAC audio stream properly muxed into an MP4 container. This means that the append
stage properly succeeds and AAC buffers with garbage data remain in memory. When playback starts and actual
decoding is done, the fdkaacdec Gstreamer decoder triggers a fatal error.

Unfortunately, this test doesn&apos;t reproduce the crash that happens downstream, but might be useful to detect
crashes and other player selection problems in the future caused by a decoding failure.

* LayoutTests/TestExpectations: Skip the test by default on all platforms, where corrupt AAC doesn&apos;t cause a playback error.
* LayoutTests/media/content/garbage-aac.mp4: Added.
* LayoutTests/media/media-source/media-source-append-play-corrupt-aac-expected.txt: Added.
* LayoutTests/media/media-source/media-source-append-play-corrupt-aac.html: Added. Test that the media element can be reused with a new MediaSource after a fatal MSE playback failure.
* LayoutTests/platform/glib/TestExpectations: Enable the test on glib platforms.

Canonical link: <a href="https://commits.webkit.org/253591@main">https://commits.webkit.org/253591@main</a>
</pre>















<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fe6a128a13e8eaa4896e8c467cb78dfad8dbbd52

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ ~~🧪 style~~](https://ews-build.webkit.org/#/builders/6/builds/86474 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/30402 "failed Failed to checkout and rebase branch from PR 2116") | [❌ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/17440 "failed Failed to checkout and rebase branch from PR 2116") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/95322 "failed Failed to checkout and rebase branch from PR 2116") | 
| | [✅ ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/64/builds/28822 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/25373 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/78634 "failed Failed to checkout and rebase branch from PR 2116") | [❌ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/90567 "failed Failed to checkout and rebase branch from PR 2116") 
| [✅ ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/19/builds/92090 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/64/builds/28822 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/73438 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/36/builds/78634 "failed Failed to checkout and rebase branch from PR 2116") | 
| | [✅ ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/64/builds/28822 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/43/builds/17440 "failed Failed to checkout and rebase branch from PR 2116") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/36/builds/78634 "failed Failed to checkout and rebase branch from PR 2116") | 
| [❌ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/78411 "Failed to checkout and rebase branch from PR 2116") | [✅ ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/66/builds/26707 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/43/builds/17440 "failed Failed to checkout and rebase branch from PR 2116") | [❌ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/72048 "failed Failed to checkout and rebase branch from PR 2116") | 
| | [✅ ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/26621 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/43/builds/17440 "failed Failed to checkout and rebase branch from PR 2116") | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/38/builds/72048 "failed Failed to checkout and rebase branch from PR 2116") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2559 "Built successfully and passed tests") | [✅ ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/63/builds/28300 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/73438 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/74828 "failed Failed to checkout and rebase branch from PR 2116") | 
| | [✅ ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/65/builds/28241 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/43/builds/17440 "failed Failed to checkout and rebase branch from PR 2116") | [❌ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/37/builds/74828 "failed Failed to checkout and rebase branch from PR 2116") | 
<!--EWS-Status-Bubble-End-->

calvaris commented on LayoutTests/media/media-source/media-source-append-play-corrupt-aac.html:
```diff
@@ -0,0 +1,46 @@
+<head>
+<title>Media source append and play corrupt aac file</title>
+<script src="../video-test.js"></script>
+<!--script src="../../resources/testharness.js"></script-->
```
You can remove this

calvaris reviewed (approved):
LGTM. Let's let a couple of days for more comments just in case. We can land on Friday CEST morning.

eric-carlson commented on Source/WebCore/platform/graphics/MediaPlayer.cpp:
```diff
@@ -1362,14 +1362,19 @@ void MediaPlayer::setPrivateBrowsingMode(bool privateBrowsingMode)
+            && (nextBestMediaEngine(m_currentMediaEngine) || nextMediaEngine(m_currentMediaEngine))
+#if ENABLE(MEDIA_SOURCE)
+            && !m_mediaSource
+#endif
```
This seems like the wrong way to fix this - won't it cause any engine registered after MSE to never be used?

eocanha replied:
As I understand it, if a MediaSource is attached to a media element, it means that only a MSE-aware player private (media engine) should be used. A non-MSE player private wouldn't know how to handle that source of data. So, if media engine trial fails for a MSE player private, why keep trying?

^^^ Answer: because of MockMSE. Now I see it.

<del>Are there multiple engines supporting MSE on Mac (so it's worth to keep trying in that platform)? We only have one in ports using GStreamer (it's not worth to keep trying in that platform).</del>

eocanha replied:
I'm just removing this piece of code and leaving the test.

eocanha committed:
[Media][MSE] Check that the media element can still be used after a decoding error
https://bugs.webkit.org/show_bug.cgi?id=242384

Reviewed by Xabier Rodriguez-Calvar.

The media source instance should be detached before destroying MSE private player. Otherwise, it can create a
crash under some circumstances that can only properly be reproduced in downstream ports with specific devices.

See: https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/890

A test to check that the player doesn't crash when an MSE playback decoding error happens has been added.
This test uses a corrupt AAC audio stream properly muxed into an MP4 container. This means that the append
stage properly succeeds and AAC buffers with garbage data remain in memory. When playback starts and actual
decoding is done, the fdkaacdec Gstreamer decoder triggers a fatal error.

Unfortunately, this test doesn't reproduce the crash that happens downstream, but might be useful to detect
crashes and other player selection problems in the future caused by a decoding failure.

* LayoutTests/TestExpectations: Skip the test by default on all platforms, where corrupt AAC doesn't cause a playback error.
* LayoutTests/media/content/garbage-aac.mp4: Added.
* LayoutTests/media/media-source/media-source-append-play-corrupt-aac-expected.txt: Added.
* LayoutTests/media/media-source/media-source-append-play-corrupt-aac.html: Added. Test that the media element can be reused with a new MediaSource after a fatal MSE playback failure.
* LayoutTests/platform/glib/TestExpectations: Enable the test on glib platforms.

Canonical link: https://commits.webkit.org/253591@main
